Validation webhook for ingress

MustafaSaber opened the pull request:
Introducing validation for skipper `Ingress` annotations

Related https://github.com/zalando/skipper/pull/2493 & https://github.com/zalando/skipper/issues/1618

MustafaSaber committed:
Validate `Ingress` annotations

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

MustafaSaber committed:
Rename validation webhook and narrow the condition

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

MustafaSaber committed:
Update skipper version for webhook binary

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

AlexanderYastrebov commented on cluster/manifests/deletions.yaml:
```diff
@@ -1,7 +1,7 @@
 pre_apply:
 {{ if ne .Cluster.ConfigItems.routegroups_validation "enabled" }}
-- name: "routegroup-admitter.teapot.zalan.do"
+- name: "skipper-admitter.teapot.zalan.do"
```
I think you'd need to have temporary unconditional deletion for `routegroup-admitter.teapot.zalan.do` due to rename.

MustafaSaber replied:
Good catch

AlexanderYastrebov replied:
Old `routegroup-admitter.teapot.zalan.do` should be put unconditionally and better into `post_apply` such that it is removed after `skipper-admitter.teapot.zalan.do` is created.

`skipper-admitter.teapot.zalan.do` should not be removed at all because `.Cluster.ConfigItems.routegroups_validation` condition only applies to the nested webhook inside of the `skipper-admitter.teapot.zalan.do`

MustafaSaber replied:
We will need also not to remove `skipper-admitter.teapot.zalan.do` in the first run since it's not there in the first place

AlexanderYastrebov commented on cluster/manifests/01-admission-control/routegroups-webhook.yaml:
```diff
@@ -1,11 +1,11 @@
+  name: "skipper-admitter.teapot.zalan.do"
   labels:
-    application: routegroups-admission-webhook
+    application: skipper-admission-webhook
```
Should we take a chance and use the same application `skipper-ingress` as we use for skipper (and `webhook` component)?

MustafaSaber replied:
I think that wouldn't be a problem.
FYI those application IDs are not registered.

MustafaSaber committed:
Use same application for skipper

add initial deletion

remove condition for webhook to work always with ingress at least

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

MustafaSaber commented on cluster/node-pools/master-default/userdata.yaml:
```diff
@@ -241,9 +241,8 @@ write_files:
             - mountPath: /etc/kubernetes/admission-controller-kubeconfig
               name: admission-controller-kubeconfig
               readOnly: true
-{{- if or (eq .Cluster.ConfigItems.routegroups_validation "provisioned") (eq .Cluster.ConfigItems.routegroups_validation "enabled") }}
```
Condition doesn't make sense if we always need ingress and sometimes routegroup

AlexanderYastrebov commented on cluster/node-pools/master-default/userdata.yaml:
```diff
@@ -491,7 +489,7 @@ write_files:
               -> setPath("/metrics")
               -> "http://127.0.0.1:8081";
-            routegroups_admission_webhook: Path("/routegroups-admission-webhook")
+            skipper_admission_webhook: Path("/skipper-admission-webhook")
```
Check out where else `/routegroups-admission-webhook` is used, I think you'd need to update there as well.

MustafaSaber committed:
fix deletion order

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

MustafaSaber committed:
fix metrics job naming and path

Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

AlexanderYastrebov commented:
:+1:

gargravarr commented:
:+1:
